action/protocol/staking/candidate_center.go

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2020 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package staking
+
+import (
+	"sort"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/iotex-core/action/protocol/staking/stakingpb"
+)
+
+type (
+	// CandidateCenter is a struct to manage the candidates
+	CandidateCenter struct {
+		nameMap     map[string]*Candidate
+		ownerMap    map[address.Address]*Candidate
+		operatorMap map[address.Address]*Candidate
+	}
+)
+
+// NewCandidateCenter creates an instance of CandidateCenter
+func NewCandidateCenter() *CandidateCenter {
+	return &CandidateCenter{
+		nameMap:     make(map[string]*Candidate),
+		ownerMap:    make(map[address.Address]*Candidate),
+		operatorMap: make(map[address.Address]*Candidate),
+	}
+}
+
+// Size returns number of candidates
+func (m CandidateCenter) Size() int {
+	return len(m.nameMap)
+}
+
+// ContainsName returns true if the map contains the candidate by name
+func (m CandidateCenter) ContainsName(name string) bool {
+	_, ok := m.nameMap[name]
+	return ok
+}
+
+// ContainsOwner returns true if the map contains the candidate by owner
+func (m CandidateCenter) ContainsOwner(owner address.Address) bool {
+	_, ok := m.ownerMap[owner]
+	return ok
+}
+
+// ContainsOperator returns true if the map contains the candidate by operator
+func (m CandidateCenter) ContainsOperator(operator address.Address) bool {
+	_, ok := m.operatorMap[operator]
+	return ok
+}
+
+// Get returns the candidate by name
+func (m CandidateCenter) Get(name string) *Candidate {
+	return m.nameMap[name]
+}
+
+// Put writes the candidate into map, overwrite an existing entry cause error
+func (m CandidateCenter) Put(d *Candidate) error {
+	if m.ContainsName(d.Name) || m.ContainsOwner(d.Owner) || m.ContainsOperator(d.Operator) {
+		return ErrAlreadyExist
+	}
+
+	m.nameMap[d.Name] = d
+	m.ownerMap[d.Owner] = d
+	m.operatorMap[d.Operator] = d
+	return nil
+}
+
+// Delete deletes the candidate by name
+func (m CandidateCenter) Delete(name string) {
+	d, ok := m.nameMap[name]
+	if !ok {
+		return
+	}
+
+	delete(m.nameMap, name)
+	delete(m.ownerMap, d.Owner)
+	delete(m.operatorMap, d.Operator)
+}
+
+// Serialize serializes a CandidateCenter to bytes
+func (m CandidateCenter) Serialize() ([]byte, error) {
+	l := make(CandidateList, 0, len(m.nameMap))
+	for _, v := range m.nameMap {
+		l = append(l, v)
+	}
+	sort.Sort(l)
+	lpb, err := l.toProto()
+	if err != nil {
+		return nil, err
+	}
+	return proto.Marshal(lpb)
+}
```

Comment:
This is a serialization for candidate list. I still think we could have ToCandidateList() and FromCandidateList() functions, instead of the confusing serialize and deserialize functions.

Reply:
Looking from caller perspective, this does serialize to []byte, and de-serialize from []byte, so i don't think the name is confusing.

it's just that it used CandidateList as an internal vehicle to perform the task

Reply:
i think you might mean to filter out like top36 candidates from all? we might need to save that for each block

we can have another func like ToConsensusDelegates() to do that